pkg/monitors/pingdomtransaction/pingdom-transaction-monitor.go

```diff
@@ -0,0 +1,386 @@
+			transactionCheck.Steps = append(transactionCheck.Steps, pingdomNew.Step{
+				Args: args,
+				Fn:   ptr.String(step.Function),
+			})
+		} else {
+			log.Info("Invalid Pingdom Step Args Provided")
+		}
+	}
+}
+
+// KubernetesInterface abstracts the clientset methods used
+type KubernetesInterface interface {
+	CoreV1() corev1.CoreV1Interface
+}
+
+// Ensure the real clientset satisfies the interface
+var _ KubernetesInterface = &kubernetes.Clientset{}
+
+// Ensure the fake clientset satisfies the interface
+var _ KubernetesInterface = &fake.Clientset{}
+
+// NewStepArgsByMap creates a new StepArgs object from a map
+func (service *PingdomTransactionMonitorService) NewStepArgsByMap(input map[string]string) *pingdomNew.StepArgs {
+	// First, marshal the map to JSON
+	jsonData, err := json.Marshal(input)
+	if err != nil {
+		log.Error(err, "Error marshalling map to JSON")
+		return nil
+	}
+	var stepArgs pingdomNew.StepArgs
+	err = json.Unmarshal(jsonData, &stepArgs)
+	if err != nil {
+		log.Error(err, "Error marshalling map to StepArgs")
+		return nil
+	}
+	// Replace secrets in the password field
+	err = replaceSecretValuesInArgs(&stepArgs, service.kubeClient, service.namespace)
+	if err != nil {
+		log.Error(err, "Error replacing secrets in step args")
+		return nil
+	}
+	return &stepArgs
+}
+
+// ReplaceSecretValuesInArgs replaces secrets in StepArgs with actual secret values from Kubernetes.
+// It expects secrets to be formatted as {secret:secret-name:key} in the args.
+// Returns an error if the secret or defined secret key cannot be retrieved.
+func replaceSecretValuesInArgs(args *pingdomNew.StepArgs, kubeClient KubernetesInterface, namespace string) error {
+	if args == nil {
+		return nil // No arguments provided
+	}
+	// Helper function to process a single string field
+	processField := func(field *string) error {
+		if field == nil {
+			return nil // Skip nil fields
+		}
+		secretName, secretKey := parseSecretTemplate(*field)
+		if secretName != "" && secretKey != "" {
+			secretValue, err := getSecretValue(kubeClient, namespace, secretName, secretKey)
+
+			if err != nil {
+				return fmt.Errorf("failed to get secret: %v", err)
+			}
+			*field = secretValue
+		}
+		return nil
+	}
+	// Process the Password and Input fields
+	if err := processField(args.Password); err != nil {
+		return err
+	}
+	if err := processField(args.Value); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// parseSecretTemplate extracts secret name and key from a string template.
+func parseSecretTemplate(content string) (secretName string, secretKey string) {
+	const secretPattern = `{secret:(.*):(.*)}`
+	re := regexp.MustCompile(secretPattern)
+
+	matches := re.FindStringSubmatch(content)
+	if len(matches) == 3 {
+		return matches[1], matches[2]
+	}
+	return "", ""
+}
+
+// getSecretValue retrieves a secret value from Kubernetes.
+func getSecretValue(kubeClient KubernetesInterface, namespace, secretName, secretKey string) (string, error) {
```

Comment:
do we need to define separate interfaces to do get calls here? simple get call on secret types with kubeclient can also give us secret object.

Reply:
The interface is required in order to mock the kubernetes client for test. see test `TestReplaceSecrets` https://github.com/karlderkaefer/IngressMonitorController/blob/c06ff6e38c6e2b875d9b91ebac20e33ac2d3bdf8/pkg/monitors/pingdomtransaction/pingdom-transaction-monitor_test.go#L112